### PR TITLE
[Cross Client Batching LeaderTimes - 3b] | AtlasDb accepts external request batcher for LeaderTime api

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
AtlasDb accepts external request batcher for LeaderTime api.

**Implementation Description (bullets)**:
Wired BatcherProviders through TransactionManagers

**Testing (What was existing testing like?  What have you done to improve it?)**:
TBD 

**Concerns (what feedback would you like?)**:

- Does the api look reasonable

**Where should we start reviewing?**:
TransactionManagers.java

**Priority (whenever / two weeks / yesterday)**:
Today

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 1]
```diff
@@ -1200,6 +1223,24 @@ private static LockAndTimestampServices getLockAndTimestampServices(
...
 
+    private static LeaderTimeGetter getLeaderTimeGetter(
+            String timelockNamespace,
+            Optional<TimeLockRequestBatcherProviders> timelockRequestBatcherProviders,
+            AtlasDbDialogueServiceProvider serviceProvider,
+            LeaderElectionReportingTimelockService namespacedConjureTimelockService) {
+        return timelockRequestBatcherProviders
+                .map(TimeLockRequestBatcherProviders::leaderTimeBatcherProvider)
+                .map(provider -> provider.getBatcher(getMultiClientTimelockServiceSupplier(serviceProvider)))
+                .<LeaderTimeGetter>map(batcher -> new NamespacedCoalescingLeaderTimeGetter(timelockNamespace, batcher))
```
Comment (Jolyon-S) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:1234]: wow, it really needs this type hint to work? Funky

File: atlasdb-config/src/main/java/com/palantir/atlasdb/config/BatcherProvider.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.config;
+
+import com.palantir.lock.client.InternalMultiClientConjureTimelockService;
+import java.util.function.Supplier;
+
+public interface BatcherProvider<T> {
+    T getBatcher(Supplier<InternalMultiClientConjureTimelockService> multiClientConjureTimelockService);
```
Comment (Jolyon-S) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/config/BatcherProvider.java:23]: Feels a bit weird - very specific arg but generic return type. I wonder what thoughts are on making this:
```java
T getBatcher(Supplier<K> service);
```
I can see pros and cons both ways. However, if you want to keep this param as is, maybe I'd be a bit more specific on the name (MultiClientBatchProvider? TimelockBatchProvider? idk).

**Review by Jolyon-S — COMMENTED** [review]
I think it mostly looks fine, but I haven't followed along with all the PRs in this chain so might be missing something.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:1234]: This broadly makes sense. As a matter of style I'd probably write this imperatively as there is a fair bit to take in, but this is fine if you prefer.

File: changelog/@unreleased/pr-5128.v2.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+type: feature
+feature:
+  description: AtlasDb can now be configured by its clients to use custom request
+    batcher for batching requests to leaderTime api on Timelock. In the absence of
+    custom batcher, we fall back to legacy leaderTime request coalescing batcher.
```
Comment (jeremyk-91) [inline changelog/@unreleased/pr-5128.v2.yml:5]: nit: probably want articles i.e. on line 3, configured by its clients to use *a* custom request batcher; line 5, in the absence of *a* custom batcher, we fall back to *a* legacy leaderTime request coalescing batcher.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 4]
```diff
@@ -244,6 +252,8 @@ boolean allSafeForLogging() {
 
     abstract UserAgent userAgent();
 
+    abstract Optional<TimeLockRequestBatcherProviders> batcherProviders();
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:255]: can we name this `timelockRequestBatcherProviders` for consistency?

**Review by jeremyk-91 — COMMENTED** [review]
🕐 0:14 🏁 0:47

Broadly looks good, couple of small suggestions.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 5]
```diff
@@ -1200,6 +1224,28 @@ private static LockAndTimestampServices getLockAndTimestampServices(
...
+            String timelockNamespace,
+            Optional<TimeLockRequestBatcherProviders> timelockRequestBatcherProviders,
+            AtlasDbDialogueServiceProvider serviceProvider,
+            LeaderElectionReportingTimelockService namespacedConjureTimelockService) {
+
+        if (!timelockRequestBatcherProviders.isPresent()) {
+            return new LegacyLeaderTimeGetter(namespacedConjureTimelockService);
+        }
+
+        LeaderTimeCoalescingBatcher batcher = timelockRequestBatcherProviders.get()
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java:1237]: go to the next line :)

**Review by jeremyk-91 — APPROVED** [review]
👍

**Pull request merged**